lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,439 @@
+  // current Op without violating other peer BrgemmCallOp in the same level; For
+  // example, one scf.ForOp contains two TILECFG in the same level, then we
+  // cannot hoist any of them.
+  // NOLINTBEGIN(performance-unnecessary-value-param)
+  void expandInvariantScopeWithCond(
+      OpStructInfoMap &structInfo, Operation *op,
+      std::function<bool(Operation *)> controlFlowAllow,
+      std::function<bool(Operation *, const OpStructInfoMap &, Operation *,
+                         const DenseSet<Operation *> &)>
+          peerAllow) {
+    // NOLINTEND(performance-unnecessary-value-param)
+    auto opIter = structInfo.find(op);
+    assert(opIter != structInfo.end());
+    auto contextRoot = opIter->second.contextRoot;
+    auto current = op;
+    auto currIter = opIter;
+    auto parent = opIter->second.parent;
+    while (parent != contextRoot) {
+      auto parentIter = structInfo.find(parent);
+      assert(parentIter != structInfo.end());
+      // Verify whether we can expand the scope to direct parent
+      bool isControlFlowAllow = controlFlowAllow(parent);
+      bool isPeerAllow =
+          peerAllow(op, structInfo, current, parentIter->second.child);
+      if (!isControlFlowAllow || !isPeerAllow) {
+        break;
+      }
+      current = parent;
+      currIter = parentIter;
+      parent = parentIter->second.parent;
+    }
+
+    opIter->second.maxInvariantScope = current;
+  }
+
+  void expandInvariantScope(OpStructInfoMap &structInfo, Operation *op) {
+    BrgemmCallType brgemmCallType = getBrgemmCallType(op);
+    assert(brgemmCallType == BrgemmCallType::TILECFG ||
+           brgemmCallType == BrgemmCallType::TILERELEASE);
+
+    if (brgemmCallType == BrgemmCallType::TILECFG) {
+      expandInvariantScopeWithCond(
+          structInfo, op,
+          [](Operation *op) -> bool {
+            return !llvm::isa<scf::IfOp>(op) &&
+                   !llvm::isa<scf::IndexSwitchOp>(op);
+          },
+          [](Operation *self, const OpStructInfoMap &structInfo,
+             Operation *current, const DenseSet<Operation *> &peers) -> bool {
+            for (auto peer : peers) {
+              if (peer == current)
+                continue;
+              if (peer == self->getOperand(0).getDefiningOp()) {
+                // Don't break operand domination
+                return false;
+              }
+              const auto iter = structInfo.find(peer);
+              assert(iter != structInfo.end());
+              const auto &containType = iter->second.containBrgemmCallType;
+              if (containType[BrgemmCallType::DISPATCH] ||
+                  containType[BrgemmCallType::TILECFG]) {
+                return false;
+              }
+            }
+            return true;
+          });
+    } else { // brgemmCallType == BrgemmCallType::TILERELEASE
+      expandInvariantScopeWithCond(
+          structInfo, op,
+          [](Operation *op) -> bool {
+            return !llvm::isa<scf::IfOp>(op) &&
```

Comment:
Suggestion: you can use `llvm::isa<IfOp, IndexSwitchOp>(op)` to check if op is any of these two types.

Reply:
Fixed.